cibuildwheel/__main__.py

```diff
@@ -156,7 +159,17 @@ def main() -> None:
     test_extras = get_option_from_environment('CIBW_TEST_EXTRAS', platform=platform, default='')
     build_verbosity_str = get_option_from_environment('CIBW_BUILD_VERBOSITY', platform=platform, default='')
 
-    build_selector = BuildSelector(build_config=build_config, skip_config=skip_config)
+    project_files = ProjectFiles(package_dir)
+
+    if not project_files.exists():
+        print('cibuildwheel: Could not find setup.py, setup.cfg or pyproject.toml at root of package', file=sys.stderr)
+        sys.exit(2)
+
+    # Passing this in as an environment variable will override pyproject.toml, setup.cfg, or setup.py
+    requires_python_str: Optional[str] = os.environ.get('CIBW_PROJECT_REQUIRES_PYTHON') or project_files.get_requires_python_str()
```

Comment:
Since we're into the style discussion, I still think this would be simpler as a function.

While the class is mostly stateless, as a reader reading this code, you can't tell it's stateless without reading the source of the class. So we'd still have to jump to that file to read what 'exists' and 'get_requires_python_str' mean.

Once we get there, we have to understand those property accessors, which reference more accessors, the `dig` function, and then finally the `get_requires_python_str` method,  just to understand what's going on.

Going back to your original solution, but factored as a function-

```python
def get_package_python_requires(package_dir: Path) -> Optional[str]:
    pyproject_toml = package_dir / 'pyproject.toml'
    try:
        info = toml.load(pyproject_toml)
        return info['project']['requires-python']
    except (FileNotFoundError, KeyError):
        pass

    # Read in from setup.cfg:options.python_requires
    setup_cfg = package_dir / 'setup.cfg'
    config = ConfigParser()
    try:
        config.read(setup_cfg)
        return config['options']['python_requires']
    except (FileNotFoundError, KeyError):
        pass
    
    # TODO: read from setup.py
```

It's just... so much clearer!

(There's also a smell in that the error message `print('cibuildwheel: Could not find setup.py, setup.cfg or pyproject.toml at root of package'` understands the mechanism of `project_files.exists()`. I'd consider that a 'leaky abstraction', though it could be solved by a custom exception, it's probably not worth it in this case.)

Reply:
Yes. I agree with @henryiii that the original class design abstract things better in a large, complex project, but part of the charm of `cibuildwheel` (at least to me) is that it's almost "just a large script" that glues things together, rather than a full OO program.

Reply:
(or at least, that's the current state and style, and it seems like that has worked and made things flexible and intuitive so far)